lib/SILOptimizer/Transforms/SimplifyCFG.cpp

```diff
@@ -1878,14 +1878,38 @@ static bool onlyForwardsNone(SILBasicBlock *noneBB, SILBasicBlock *someBB,
 ///    \            ... (more bbs?)
 ///     \           /
 ///       ulimateBB
-static bool hasSameUlitmateSuccessor(SILBasicBlock *noneBB, SILBasicBlock *someBB) {
+static bool hasSameUltimateSuccessor(SILBasicBlock *noneBB, SILBasicBlock *someBB) {
+  // Make sure that both our some, none blocks both have single successors that
+  // are not themselves (which can happen due to single block loops).
   auto *someSuccessorBB = someBB->getSingleSuccessorBlock();
-  if (!someSuccessorBB)
+  if (!someSuccessorBB || someSuccessorBB == someBB)
     return false;
   auto *noneSuccessorBB = noneBB->getSingleSuccessorBlock();
-  while (noneSuccessorBB != nullptr && noneSuccessorBB != someSuccessorBB)
-    noneSuccessorBB = noneSuccessorBB->getSingleSuccessorBlock();
-  return noneSuccessorBB == someSuccessorBB;
+  if (!noneSuccessorBB || noneSuccessorBB == noneBB)
+    return false;
+
+  // If we immediately find a diamond, return true. We are done.
+  if (noneSuccessorBB == someSuccessorBB)
```

Comment:
This check is redundant. It will be handled by the while-condition below.

Reply:
While it is redundant, the idea was to have an early exit so that the reader of the code wouldn't need to process the while loop to look at this simple case.

That being said, I do not have a problem with eliminating this if it /really/ bugs you. I don't think it hurts and I think it makes the code easier to understand.